cosim/black-parrot-example/vivado/tcl/constraints.tcl

```diff
@@ -0,0 +1,98 @@
+
+# In BP, there is a 4-1 clock multiplexer that selects between 3 clocks and 1'b0 as the mtime clock.
+# (The 4-1 clock mux)
+#
+#
+#        |\
+#    +---| \ (rtc_mux1)
+#        | |-+
+#    +---| / |  |\
+#        |/  +--| \ (rtc_mux3)
+#               | |---+
+#        |\  +--| /
+#    +---| \ |  |/
+#        | |-+
+#    +---| / (rtc_mux2)
+#        |/
+#
+# 'join' converts a list into a string
+set rtc_mux1 [join [get_cells -hier *bufgmux_ctrl1]]
+set rtc_mux2 [join [get_cells -hier *bufgmux_ctrl2]]
+set rtc_mux3 [join [get_cells -hier *bufgmux_ctrl3]]
+#
+# Ordinarily, we would create 2 generated clocks for rtc_mux1, 2 for rtc_mux2 and 4 for rtc_mux3 to
+# consider all the 3 clocks during STA. For example, it could be something like
+#
+#   create_generated_clock -name rtc_mux1_src1 -divide_by 1 -add -master_clock clk1 \
+#       -source [get_pins rtc_mux1/I0] [get_pins rtc_mux1/O]
+#   create_generated_clock -name rtc_mux1_src2 -divide_by 1 -add -master_clock clk2 \
+#       -source [get_pins rtc_mux1/I1] [get_pins rtc_mux1/O]
+#   set_clock_groups -physically_exclusive -group rtc_mux1_src1 -group rtc_mux1_src2
+#
+#   create_generated_clock -name rtc_mux2_src1 -divide_by 1 -add -master_clock clk3 \
+#       -source [get_pins rtc_mux2/I0] [get_pins rtc_mux2/O]
+#   create_generated_clock -name rtc_mux2_src2 -divide_by 1 -add -master_clock clk4 \
+#       -source [get_pins rtc_mux2/I1] [get_pins rtc_mux2/O]
+#   set_clock_groups -physically_exclusive -group rtc_mux2_src1 -group rtc_mux2_src2
+#
+#   create_generated_clock -name rtc_mux3_src1 -divide_by 1 -add -master_clock rtc_mux1_src1 \
+#       -source [get_pins rtc_mux3/I0] [get_pins rtc_mux3/O]
+#   create_generated_clock -name rtc_mux3_src2 -divide_by 1 -add -master_clock rtc_mux1_src2 \
+#       -source [get_pins rtc_mux3/I0] [get_pins rtc_mux3/O]
+#   create_generated_clock -name rtc_mux3_src3 -divide_by 1 -add -master_clock rtc_mux2_src1 \
+#       -source [get_pins rtc_mux3/I1] [get_pins rtc_mux3/O]
+#   create_generated_clock -name rtc_mux3_src4 -divide_by 1 -add -master_clock rtc_mux2_src2 \
+#       -source [get_pins rtc_mux3/I1] [get_pins rtc_mux3/O]
+#   set_clock_groups -physically_exclusive -group rtc_mux3_src1 -group rtc_mux3_src2 -group rtc_mux3_src3 -group rtc_mux3_src4
+#
+# However, since
+#   1. the w_reset_i pin of mtime_gray is wired to reset_i
+#   2. the select for the 4-1 mux is initialized to 2'b00 when reset_i is high
+#   3. clk_i is always faster than rt_clk_i
+# , we should only consider clk_i for mtime clock. In order to do this, instead of creating all the generated clocks like
+# the one in the above example, we only create necessary clocks for clk_i:
+# clk_i running through rtc_mux1
+create_generated_clock -name rtc_mux1_clk1 -divide_by 1 -source [get_pins $rtc_mux1/I0] -add -master_clock [get_clocks clk_fpga_0] [get_pins $rtc_mux1/O]
```

Comment:
Only the downsampled generated clock should be necessary -- Vivado should propagate the rest through the muxes correctly

Reply:
case analysis 00 on the mux will also correcty constrain the reset